File: app/src/main/java/com/example/sarahn/waterreminderapp/Utils/SharedPrefUtils.java
package com.example.sarahn.waterreminderapp.Utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.widget.Toast;

/**
 * Created by SarahN on 6/6/2017.
 */
public class SharedPrefUtils {

    private final static String KEY_IS_TRUE = "is_launch_first_time";
    private final static String KEY_WEIGHT = "weight";
    private final static String KEY_ACTIVITY_LEVEL = "activity_level";
    private final static String KEY_CLIMATE = "climate";
    private static int[] userData;
    private final static int DEFAULT_VALUE = 0;
    private final static String DEFUALT_STRING_VALUE = null;
    private final static String KEY_CONSUMED = "consumed";
    private final static String KEY_REMAINED = "remained";
    private final static String KEY_REQUIRED = "required";

    private final static String KEY_START_HOUR = "start_hour";
    private final static String KEY_END_HOUR = "end_hour";
    private final static String KEY_START_MIN = "start_min";
    private final static String KEY_END_MIN = "end_min";

    private final static String KEY_ISBOOT = "is_boot";

    private final static String KEY_DURATION = "duration";


//    public SharedPrefUtils(Context context) {
//        this.context = context;
//        sharedPreferences = this.context.getSharedPreferences(SHARED_PREFERENCE_NAME, 0);
//        editor = sharedPreferences.edit();
//    }

    synchronized public static void setIsTrue(Boolean istrue, Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_IS_TRUE);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putBoolean(KEY_IS_TRUE, istrue);
        editor.apply();
    }

    public static Boolean getIsTrue(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getBoolean(KEY_IS_TRUE, true);
    }

//    synchronized public static void setUserData(Context context, int weight, int activity, int climate){
//        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
//        SharedPreferences.Editor editor = prefs.edit();
//        editor.putInt(KEY_WEIGHT,weight);
//        editor.putInt(KEY_ACTIVITY_LEVEL, activity);
//        editor.putInt(KEY_CLIMATE, climate);
//        editor.apply();
//    }
    synchronized public static void setWeight(Context context, int weight){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_WEIGHT);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(KEY_WEIGHT, weight);
        editor.apply();
    }

    public static int getWeight(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getInt(KEY_WEIGHT, DEFAULT_VALUE);
    }

    public static String getActivityLevel(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getString(KEY_ACTIVITY_LEVEL, DEFUALT_STRING_VALUE);
    }

    public static String getClimate(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getString(KEY_CLIMATE, DEFUALT_STRING_VALUE);
    }

    synchronized public static void setActivityLevel(Context context,String activity){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_ACTIVITY_LEVEL);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(KEY_ACTIVITY_LEVEL, activity);
        editor.apply();

    }

    synchronized public static void setClimate(Context context, String climate){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_CLIMATE);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(KEY_CLIMATE, climate);
        editor.apply();

    }

    synchronized public static void setConsumed(Context context, int consume){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_CONSUMED);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(KEY_CONSUMED, consume);
        editor.apply();
    }

    public static int getConsumed(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getInt(KEY_CONSUMED, DEFAULT_VALUE);
    }

    synchronized public static void setRemained(Context context, int consume){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
            prefs.edit().remove(KEY_REMAINED);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(KEY_REMAINED, consume);
        editor.apply();
    }

    public static int getRemained(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getInt(KEY_REMAINED, DEFAULT_VALUE);
    }

    synchronized public static void setRequired(Context context, int consume){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_REQUIRED);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(KEY_REQUIRED, consume);
        editor.apply();
    }

    synchronized public static void setDuration(Context context, int duration){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_DURATION);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(KEY_DURATION, duration);
        editor.apply();
    }

    public static int getDuration(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getInt(KEY_DURATION, DEFAULT_VALUE);
    }

    public static void removeConsumed(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        //changing from commit to apply
        prefs.edit().remove(KEY_CONSUMED).apply();
        prefs.edit().remove(KEY_REMAINED).apply();
        Logging.logMessage("removed consumed ");
        Logging.logMessage("clear---" + getConsumed(context));
        Logging.logMessage("clear remained ---" + getRemained(context));
    }

    public static boolean getIsBoot(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getBoolean(KEY_ISBOOT, false);
    }

    synchronized public static void setIsBoot(Context context, boolean boot){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_ISBOOT);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putBoolean(KEY_ISBOOT, boot);
        editor.apply();
    }

    public static int getRequired(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getInt(KEY_REQUIRED, DEFAULT_VALUE);
    }

    synchronized public static void setStartHour(Context context, int consume){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_START_HOUR);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(KEY_START_HOUR, consume);
        editor.apply();
    }

    public static int getStartHour(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getInt(KEY_START_HOUR, DEFAULT_VALUE);
    }
    synchronized public static void setEndHour(Context context, int consume){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_END_HOUR);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(KEY_END_HOUR, consume);
        editor.apply();
    }

    public static int getEndHour(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getInt(KEY_END_HOUR, DEFAULT_VALUE);
    }
    synchronized public static void setStartMin(Context context, int consume){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_START_MIN);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(KEY_START_MIN, consume);
        editor.apply();
    }

    public static int getStartMin(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getInt(KEY_START_MIN, DEFAULT_VALUE);
    }
    synchronized public static void setEndMin(Context context, int consume){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        prefs.edit().remove(KEY_END_MIN);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(KEY_END_MIN, consume);
        editor.apply();
    }

    public static int getEndMin(Context context){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getInt(KEY_END_MIN, DEFAULT_VALUE);
    }





//    synchronized public static void setRequired(Context context, int climate){
//        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
//        prefs.edit().remove(KEY_CLIMATE);
//        SharedPreferences.Editor editor = prefs.edit();
//        editor.putString(, climate);
//        editor.apply();
//
//    }
//
//    public static int getRequired(){
//
//    }

//    public static int[] getUserData(Context context){
//        // how to make lines of code follow synchronization, step by step execution
//        SharedPreferences prefs =  PreferenceManager.getDefaultSharedPreferences(context);
//        int weight = prefs.getInt(KEY_WEIGHT, DEFAULT_VALUE);
//        int activity = prefs.getInt(KEY_ACTIVITY_LEVEL, DEFAULT_VALUE);
//        int climate = prefs.getInt(KEY_CLIMATE, DEFAULT_VALUE);
//        userData = new int[]{weight, activity, climate};
//
//        return userData;
//    }
}

File: app/src/main/java/com/example/sarahn/waterreminderapp/Utils/SharedPrefUtils.kt
package com.example.sarahn.waterreminderapp.Utils

import android.content.Context
import android.preference.PreferenceManager
import com.example.sarahn.waterreminderapp.Utils.Logging.logMessage

/**
 * Created by SarahN on 6/6/2017.
 */
object SharedPrefUtils {
    private const val KEY_IS_TRUE = "is_launch_first_time"
    private const val KEY_WEIGHT = "weight"
    private const val KEY_ACTIVITY_LEVEL = "activity_level"
    private const val KEY_CLIMATE = "climate"
    private val userData: IntArray? = null
    private const val DEFAULT_VALUE = 0
    private val DEFUALT_STRING_VALUE: String? = null
    private const val KEY_CONSUMED = "consumed"
    private const val KEY_REMAINED = "remained"
    private const val KEY_REQUIRED = "required"
    private const val KEY_START_HOUR = "start_hour"
    private const val KEY_END_HOUR = "end_hour"
    private const val KEY_START_MIN = "start_min"
    private const val KEY_END_MIN = "end_min"
    private const val KEY_ISBOOT = "is_boot"
    private const val KEY_DURATION = "duration"

    //    public SharedPrefUtils(Context context) {
    //        this.context = context;
    //        sharedPreferences = this.context.getSharedPreferences(SHARED_PREFERENCE_NAME, 0);
    //        editor = sharedPreferences.edit();
    //    }
    @JvmStatic
    @Synchronized
    fun setIsTrue(istrue: Boolean?, context: Context?) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_IS_TRUE)
        val editor = prefs.edit()
        editor.putBoolean(KEY_IS_TRUE, istrue!!)
        editor.apply()
    }

    fun getIsTrue(context: Context?): Boolean {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getBoolean(KEY_IS_TRUE, true)
    }

    //    synchronized public static void setUserData(Context context, int weight, int activity, int climate){
    //        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
    //        SharedPreferences.Editor editor = prefs.edit();
    //        editor.putInt(KEY_WEIGHT,weight);
    //        editor.putInt(KEY_ACTIVITY_LEVEL, activity);
    //        editor.putInt(KEY_CLIMATE, climate);
    //        editor.apply();
    //    }
    @Synchronized
    fun setWeight(context: Context?, weight: Int) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_WEIGHT)
        val editor = prefs.edit()
        editor.putInt(KEY_WEIGHT, weight)
        editor.apply()
    }

    fun getWeight(context: Context?): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getInt(KEY_WEIGHT, DEFAULT_VALUE)
    }

    fun getActivityLevel(context: Context?): String? {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getString(KEY_ACTIVITY_LEVEL, DEFUALT_STRING_VALUE)
    }

    fun getClimate(context: Context?): String? {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getString(KEY_CLIMATE, DEFUALT_STRING_VALUE)
    }

    @Synchronized
    fun setActivityLevel(context: Context?, activity: String?) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_ACTIVITY_LEVEL)
        val editor = prefs.edit()
        editor.putString(KEY_ACTIVITY_LEVEL, activity)
        editor.apply()
    }

    @Synchronized
    fun setClimate(context: Context?, climate: String?) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_CLIMATE)
        val editor = prefs.edit()
        editor.putString(KEY_CLIMATE, climate)
        editor.apply()
    }

    @Synchronized
    fun setConsumed(context: Context?, consume: Int) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_CONSUMED)
        val editor = prefs.edit()
        editor.putInt(KEY_CONSUMED, consume)
        editor.apply()
    }

    fun getConsumed(context: Context?): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getInt(KEY_CONSUMED, DEFAULT_VALUE)
    }

    @Synchronized
    fun setRemained(context: Context?, consume: Int) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_REMAINED)
        val editor = prefs.edit()
        editor.putInt(KEY_REMAINED, consume)
        editor.apply()
    }

    fun getRemained(context: Context?): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getInt(KEY_REMAINED, DEFAULT_VALUE)
    }

    @Synchronized
    fun setRequired(context: Context?, consume: Int) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_REQUIRED)
        val editor = prefs.edit()
        editor.putInt(KEY_REQUIRED, consume)
        editor.apply()
    }

    @Synchronized
    fun setDuration(context: Context?, duration: Int) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_DURATION)
        val editor = prefs.edit()
        editor.putInt(KEY_DURATION, duration)
        editor.apply()
    }

    fun getDuration(context: Context?): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getInt(KEY_DURATION, DEFAULT_VALUE)
    }

    fun removeConsumed(context: Context?) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        //changing from commit to apply
        prefs.edit().remove(KEY_CONSUMED).apply()
        prefs.edit().remove(KEY_REMAINED).apply()
        logMessage("removed consumed ")
        logMessage("clear---" + getConsumed(context))
        logMessage("clear remained ---" + getRemained(context))
    }

    fun getIsBoot(context: Context?): Boolean {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getBoolean(KEY_ISBOOT, false)
    }

    @Synchronized
    fun setIsBoot(context: Context?, boot: Boolean) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_ISBOOT)
        val editor = prefs.edit()
        editor.putBoolean(KEY_ISBOOT, boot)
        editor.apply()
    }

    fun getRequired(context: Context?): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getInt(KEY_REQUIRED, DEFAULT_VALUE)
    }

    @Synchronized
    fun setStartHour(context: Context?, consume: Int) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_START_HOUR)
        val editor = prefs.edit()
        editor.putInt(KEY_START_HOUR, consume)
        editor.apply()
    }

    fun getStartHour(context: Context?): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getInt(KEY_START_HOUR, DEFAULT_VALUE)
    }

    @Synchronized
    fun setEndHour(context: Context?, consume: Int) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_END_HOUR)
        val editor = prefs.edit()
        editor.putInt(KEY_END_HOUR, consume)
        editor.apply()
    }

    fun getEndHour(context: Context?): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getInt(KEY_END_HOUR, DEFAULT_VALUE)
    }

    @Synchronized
    fun setStartMin(context: Context?, consume: Int) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_START_MIN)
        val editor = prefs.edit()
        editor.putInt(KEY_START_MIN, consume)
        editor.apply()
    }

    fun getStartMin(context: Context?): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getInt(KEY_START_MIN, DEFAULT_VALUE)
    }

    @Synchronized
    fun setEndMin(context: Context?, consume: Int) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        prefs.edit().remove(KEY_END_MIN)
        val editor = prefs.edit()
        editor.putInt(KEY_END_MIN, consume)
        editor.apply()
    }

    fun getEndMin(context: Context?): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        return prefs.getInt(KEY_END_MIN, DEFAULT_VALUE)
    } //    synchronized public static void setRequired(Context context, int climate){
    //        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
    //        prefs.edit().remove(KEY_CLIMATE);
    //        SharedPreferences.Editor editor = prefs.edit();
    //        editor.putString(, climate);
    //        editor.apply();
    //
    //    }
    //
    //    public static int getRequired(){
    //
    //    }
    //    public static int[] getUserData(Context context){
    //        // how to make lines of code follow synchronization, step by step execution
    //        SharedPreferences prefs =  PreferenceManager.getDefaultSharedPreferences(context);
    //        int weight = prefs.getInt(KEY_WEIGHT, DEFAULT_VALUE);
    //        int activity = prefs.getInt(KEY_ACTIVITY_LEVEL, DEFAULT_VALUE);
    //        int climate = prefs.getInt(KEY_CLIMATE, DEFAULT_VALUE);
    //        userData = new int[]{weight, activity, climate};
    //
    //        return userData;
    //    }
}